src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -307,16 +312,33 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
             return newSolution;
         }
 
-        static Solution UpdateAddedDocumentToExistingContentsInSolution(Solution solution, DocumentId addedDocumentId)
+        static (Solution newSolution, ProjectId? relatedProjectId) UpdateAddedDocumentToExistingContentsInSolution(
+            Solution solution, DocumentId addedDocumentId, ProjectId? relatedProjectIdHint)
         {
-            var relatedDocumentIds = solution.GetRelatedDocumentIds(addedDocumentId);
-            foreach (var relatedDocumentId in relatedDocumentIds)
-            {
-                var relatedDocument = solution.GetRequiredDocument(relatedDocumentId);
-                return solution.WithDocumentContentsFrom(addedDocumentId, relatedDocument.DocumentState, forceEvenIfTreesWouldDiffer: false);
-            }
+            Contract.ThrowIfTrue(addedDocumentId.ProjectId == relatedProjectIdHint);
 
-            return solution;
+            // Look for a related document we can create our contents from.  We only have to look for a single related
+            // doc as we'll be done once we update our contents to theirs.  Note: GetFirstRelatedDocumentId will also
+            // not search the project that addedDocumentId came from.  So this will help ensure we don't repeatedly add
+            // documents to a project, then look for related docs *within that project*, forcing the file-path map in it
+            // to be recreated for each document.
+            var relatedDocumentId = solution.GetFirstRelatedDocumentId(addedDocumentId, relatedProjectIdHint);
+
+            // Should never return a file as its own related document
+            Contract.ThrowIfTrue(relatedDocumentId == addedDocumentId);
```

Comment:
>Contract.ThrowIfTrue(relatedDocumentId == addedDocumentId); [](http://example.com/codeflow?start=12&length=59)

nit: I think this would fit better after the null check below

Reply:
done.